mobile-academy/src/main/java/org/motechproject/nms/mobileacademy/service/impl/SmsNotificationServiceImpl.java

```diff
@@ -0,0 +1,162 @@
+import org.motechproject.nms.mobileacademy.service.SmsNotificationService;
+import org.motechproject.server.config.SettingsFacade;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Service;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.StringWriter;
+import java.io.UnsupportedEncodingException;
+
+/**
+ * This handles all the integration pieces between MA and sms module to trigger and handle notifications
+ * for course completion
+ */
+@Service("smsNotificationService")
+public class SmsNotificationServiceImpl implements SmsNotificationService {
+
+    private static final String COURSE_COMPLETED = "nms.ma.course.completed";
+
+    private static final String SMS_STATUS = "nms.ma.sms.deliveryStatus";
+
+    private static final String SMS_NOTIFICATION_URL = "imi.sms.notification.url";
+
+    private static final String SMS_MESSAGE_CONTENT = "imi.sms.course.completion.message";
+
+    private static final String CALLBACK_URL = "imi.sms.status.callback.url";
+
+    private static final String SMS_SENDER_ID = "imi.sms.sender.id";
+
+    private static final String SMS_TEMPLATE_FILE = "smsTemplate.json";
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SmsNotificationServiceImpl.class);
+
+    private CompletionRecordDataService completionRecordDataService;
+
+    private SettingsFacade settingsFacade;
+
+    @Autowired
+    public SmsNotificationServiceImpl(CompletionRecordDataService completionRecordDataService,
+                                      @Qualifier("maImiSettings") SettingsFacade settingsFacade) {
+        this.completionRecordDataService = completionRecordDataService;
+        this.settingsFacade = settingsFacade;
+    }
+
+    @MotechListener(subjects = { COURSE_COMPLETED })
+    public void sendSmsNotification(MotechEvent event) {
+
+        LOGGER.debug("Handling course completion notification event");
+        Long callingNumber = (Long) event.getParameters().get("callingNumber");
+        CompletionRecord cr = completionRecordDataService.findRecordByCallingNumber(callingNumber);
+
+        if (cr == null) {
+            // this should never be possible since the event dispatcher upstream adds the record
+            LOGGER.error("No completion record found for callingNumber: " + callingNumber);
+        }
+
+        cr.setSentNotification(sendNotificationRequest(callingNumber));
+        completionRecordDataService.update(cr);
+    }
+
+    @MotechListener(subjects = { SMS_STATUS })
+    public void updateSmsStatus(MotechEvent event) {
+        LOGGER.debug("Handling update sms delivery status event");
+
+        String callingNumber = (String) event.getParameters().get("address");
+        int startIndex = callingNumber.indexOf(':') + 2;
+        callingNumber = callingNumber.substring(startIndex);
```

Comment:
What is the format of calling number?  Should this be done when the post from IMI comes and if it isn't correct issue a bad request response?